cirq-core/cirq/transformers/routing/line_initial_mapper.py

```diff
@@ -0,0 +1,237 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Concrete implementation of AbstractInitialMapper that places lines of qubits onto the device."""
+
+from typing import Dict, Optional, TYPE_CHECKING
+import networkx as nx
+
+from cirq import value, _compat
+from cirq.transformers import routing
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class LineInitialMapper(routing.AbstractInitialMapper):
+    """Places logical qubits in the circuit onto physical qubits on the device."""
+
+    def __init__(self, device_graph: nx.Graph) -> None:
+        """Initializes a LineInitialMapper.
+
+        Args:
+            device_graph: device graph
+        """
+        # TODO: Tanuj, should this logic be done instead at the beginning of the routing transformer
+        # so that we don't have to repeat it for each InitialMapper and for the MappingManager?
+        if nx.is_directed(device_graph):
+            self.device_graph = nx.DiGraph()
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(sorted(list(device_graph.edges)))
+        else:
+            self.device_graph = nx.Graph()
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(
+                sorted(list(sorted(edge) for edge in device_graph.edges))
+            )
+
+    def _make_circuit_graph(self, circuit: 'cirq.AbstractCircuit') -> nx.Graph:
```

Comment:
Consider returning the graph as a `List[List[cirq.Qid]]` (i.e. a list of paths) instead of an `nx.Graph`; since that's what's needed in the `initial_mapping` function. It'll help avoid conversion to/from networkx representation to paths representation. It would also give you the "sort by largest connected component" by just sorting the outer list based on size of inner list and "iterating from one end point to other" would just correspond to iterating on the inner list. This would also get rid of the `next_logical` method.